nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/CustomKerberosLogin.java

```diff
@@ -0,0 +1,245 @@
+ * to avoid making system calls to kinit when the ticket cache is being used, and to avoid exiting the refresh thread so that
+ * it may recover if the ticket cache is externally refreshed.
+ *
+ * The re-login thread follows a similar approach used by NiFi's KerberosUser which attempts to call tgt.refresh()
+ * and falls back to a logout/login.
+ *
+ * The Kafka client is configured to use this login by setting SaslConfigs.SASL_LOGIN_CLASS in {@link KafkaProcessorUtils}
+ * when the SASL mechanism is GSSAPI.
+ */
+public class CustomKerberosLogin extends AbstractLogin {
+    private static final Logger log = LoggerFactory.getLogger(CustomKerberosLogin.class);
+
+    private Thread t;
+    private boolean isKrbTicket;
+
+    private String principal;
+
+    private double ticketRenewWindowFactor;
+    private long minTimeBeforeRelogin;
+
+    private volatile Subject subject;
+
+    private LoginContext loginContext;
+    private String serviceName;
+
+    @Override
+    public void configure(Map<String, ?> configs, String contextName, Configuration configuration,
+                          AuthenticateCallbackHandler callbackHandler) {
+        super.configure(configs, contextName, configuration, callbackHandler);
+        this.ticketRenewWindowFactor = (Double) configs.get(SaslConfigs.SASL_KERBEROS_TICKET_RENEW_WINDOW_FACTOR);
+        this.minTimeBeforeRelogin = (Long) configs.get(SaslConfigs.SASL_KERBEROS_MIN_TIME_BEFORE_RELOGIN);
+        this.serviceName = getServiceName(configs, contextName, configuration);
+    }
+
+    /**
+     * Performs login for each login module specified for the login context of this instance and starts the thread used
+     * to periodically re-login to the Kerberos Ticket Granting Server.
+     */
+    @Override
+    public LoginContext login() throws LoginException {
+        loginContext = super.login();
+        subject = loginContext.getSubject();
+        isKrbTicket = !subject.getPrivateCredentials(KerberosTicket.class).isEmpty();
+
+        AppConfigurationEntry[] entries = configuration().getAppConfigurationEntry(contextName());
+        if (entries.length == 0) {
+            principal = null;
+        } else {
+            // there will only be a single entry
+            AppConfigurationEntry entry = entries[0];
+            if (entry.getOptions().get("principal") != null)
+                principal = (String) entry.getOptions().get("principal");
+            else
+                principal = null;
+        }
+
+        if (!isKrbTicket) {
+            log.debug("[Principal={}]: It is not a Kerberos ticket", principal);
+            t = null;
+            // if no TGT, do not bother with ticket management.
+            return loginContext;
+        }
+        log.debug("[Principal={}]: It is a Kerberos ticket", principal);
+
+        t = KafkaThread.daemon(String.format("kafka-kerberos-refresh-thread-%s", principal), () -> {
```

Comment:
Is the `login()` method called only once in the lifecycle of this class? This line overwrites the previous Thread value without interrupting the thread. In addition, it might be helpful to refactor this approach to an explicit `Runnable`, and then use an `ExecutorService`, as opposed to creating a new Thread.

Reply:
This class is a copy/paste with modifications from the code provided by Kafka:

https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/security/kerberos/KerberosLogin.java

Since we've actually been using the Kafka version this whole time, I think we are safe with login only being called once, although I agree it could protect against overwriting the thread. 

I thought about making bigger refactorings like the Runnable or ExecutorService, but I hesitated because I thought there might be some value in keeping a similar structure/approach to the original Kafka code for the case that we may want to compare future changes made to their version to see how it may apply to ours.

Reply:
Thanks for making the adjustments and providing that reference link. Given the existing usage, it makes sense for this NiFi implementation to follow the same basic pattern as the Kafka class for now.